### PR TITLE
Add operators `set-attributes` and `chart`

### PR DESCRIPTION
**Pull request by eliaskosunen**
Resolves https://github.com/tenzir/issues/issues/957, previous PR #3601 

This PR introduces two new operators, `set-attributes`, and `chart`. To avoid copy-pasting code, these are implemented with a helper class, `detail::set_attributes_operator_helper`.

`set-attributes [key=value]...` takes in input, and sets the requested attributes. No input validation is performed.

```
$ tenzir 'show version | set-attributes foo=bar | serve ...'
$ curl ... | jq .
{
  "next_continuation_token": "",
  "events": [
    {
      "schema_id": "28246c25e2734df4",
      "data": {
        "version": "v4.3.0-211-g863b54c2bf"
      }
    }
  ],
  "schemas": [
    {
      "schema_id": "28246c25e2734df4",
      "definition": {
        "tenzir.version": {
          "type": {
            "record": [
              {
                "version": "string"
              }
            ]
          },
          "attributes": {
            "foo": "bar"
          }
        }
      }
    }
  ]
}
```

The `chart` operator is a more specialized version of `set-attributes`. It performs validation on both the arguments it's given, and the input it receives while running.

Its syntax is `chart <type> [key=value]...`, where `<type>` is the chart type (`stacked-area`, `bar`, or `donut`), which maps to the `type` attribute. It's the only required argument, in addition to any arguments required by the chart type.

```
$ echo '{ "label": "foo", "value": 3.14 }' | \
  tenzir 'read json | chart donut label=label value=value | serve ...'
$ curl ... | jq .
{
  "next_continuation_token": "",
  "events": [
    {
      "schema_id": "95a72c3d60e175cb",
      "data": {
        "label": "good",
        "value": 3.14
      }
    },
    {
      "schema_id": "95a72c3d60e175cb",
      "data": {
        "label": "bad",
        "value": 42.0
      }
    }
  ],
  "schemas": [
    {
      "schema_id": "95a72c3d60e175cb",
      "definition": {
        "tenzir.json": {
          "type": {
            "record": [
              {
                "label": "string"
              },
              {
                "value": "double"
              }
            ]
          },
          "attributes": {
            "description": "foo bar",
            "label": "label",
            "type": "donut",
            "value": "value"
          }
        }
      }
    }
  ]
}
```

TODO:
- [x] Do we want `set-attributes` and `chart` to do this schema-wrapping thing, or set the attributes in the original schema? The former is more clear-cut, whereas the latter would need to be smart about already-existing attributes.
- [x] Figure out how to test this. The current integration tests can't inspect schema metadata. The only way to actually see the changes is through a `serve -> curl` dance, or on the frontend.

### CODE REVIEW

**Comment by dominiklohmann** [comment]
> Do we want `set-attributes` and `chart` to do this schema-wrapping thing, or set the attributes in the original schema? The former is more clear-cut, whereas the latter would need to be smart about already-existing attributes.

Our type system merges attributes when types are not named:

```cpp
auto t1 = type{"t1", string_type{}, {{"one", "1"}, {"two", "2"}}};
auto t2 = type{t1{}, {{"two", "zwei"}}};

auto t3 = type{"t1", string_type{}, {{"one", "1"}, {"two", "zwei"}}};

assert(t2 == t3); // this is guaranteed to be true:
                  // t2 overrides the attribute "two" from t1
```

If you do not name the new schema `tenzir.chart` (and I don't think we want to, as the schema name holds some semantic meaning), then you can actually just modify the schema over and over again. E.g., given `set-attributes foo=bar, baz=qux, ...` you can do this:

```cpp
auto schema = slice.schema();
for (auto [key, value] : new_attributes) {
  schema = type{schema, {{key, value}}};
}
cast(slice, schema);
```

Which probably makes it so you do not need the new helper class at all.

**Comment by eliaskosunen** [comment]
I didn't remove the helper class, since there's still quite a clear common denominator with the parsing and attribute handling logic. The helper class could be removed if I defined both of the operators in the same file, but I don't really like that idea.

Thanks for the pointer about the schema casting, I adapted the logic somewhat to align with that, and I think it made it a little cleaner.

**Review by jachris — APPROVED** [review]
Accepting this even though there is an ongoing syntax discussion. Both operators should be considered unstable and mostly for internal use right now (that's also why we didn't want documentation here).

I tested this with a `get-attributes` operator. You might want to consider to use this as well so that we have some tests, but given that everything here is experimental I don't mind too much.

#### `libtenzir/builtins/operators/get_attributes.cpp`

~~~cpp
//    _   _____   __________
//   | | / / _ | / __/_  __/     Visibility
//   | |/ / __ |_\ \  / /          Across
//   |___/_/ |_/___/ /_/       Space and Time
//
// SPDX-FileCopyrightText: (c) 2023 The Tenzir Contributors
// SPDX-License-Identifier: BSD-3-Clause

#include <tenzir/argument_parser.hpp>
#include <tenzir/pipeline.hpp>
#include <tenzir/plugin.hpp>
#include <tenzir/series_builder.hpp>

namespace tenzir::plugins::get_attributes {

namespace {

class get_attributes_operator final
  : public crtp_operator<get_attributes_operator> {
public:
  auto name() const -> std::string override {
    return "get-attributes";
  }

  auto optimize(const expression& filter, event_order order) const
    -> optimize_result override {
    (void)filter;
    return optimize_result::order_invariant(*this, order);
  }

  auto operator()(generator<table_slice> input) const
    -> generator<table_slice> {
    for (auto&& slice : input) {
      if (slice.rows() == 0) {
        co_yield {};
        continue;
      }
      auto b = series_builder{};
      auto r = b.record();
      for (auto&& [name, value] : slice.schema().attributes()) {
        r.field(name).data(value);
      }
      for (auto&& out : b.finish_as_table_slice("tenzir.attributes")) {
        co_yield std::move(out);
      }
    }
  }

  friend auto inspect(auto& f, get_attributes_operator& x) -> bool {
    return f.object(x).fields();
  }
};

class plugin final : public virtual operator_plugin<get_attributes_operator> {
public:
  auto signature() const -> operator_signature override {
    return {.transformation = true};
  }

  auto parse_operator(parser_interface& p) const -> operator_ptr override {
    auto parser = argument_parser{"get-attributes"};
    parser.parse(p);
    return std::make_unique<get_attributes_operator>();
  }
};

} // namespace

} // namespace tenzir::plugins::get_attributes

TENZIR_REGISTER_PLUGIN(tenzir::plugins::get_attributes::plugin)

~~~

File: libtenzir/src/detail/set_attributes_operator_helper.cpp [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+                          std::make_move_iterator(parsed_arguments.begin()),
+                          std::make_move_iterator(parsed_arguments.end()));
+  if (auto err = verify(cfg_); err)
+    return {std::string_view{f, l}, err};
+  return {std::string_view{f, l}, caf::error{}};
+}
+
+auto set_attributes_operator_helper::process(table_slice&& slice,
+                                             process_verify verify) const
+  -> std::pair<table_slice, caf::error> {
```
[the remark this answers is not in the document]
Comment (eliaskosunen) [reply libtenzir/src/detail/set_attributes_operator_helper.cpp:74]: None really any more. Good point.

**Comment by eliaskosunen** [comment]
I added `get-attributes` with the implementation you provided, and some integration tests utilizing it.

File: libtenzir/src/detail/set_attributes_operator_helper.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (eliaskosunen) [reply libtenzir/src/detail/set_attributes_operator_helper.cpp:75]: Ehh, it makes the client code a little nicer:

```cpp
// current
auto [result, err] = helper_.process(...);
if (!result)
  ctrl.warn(std::move(err));
co_yield result;

// with expected
auto result = helper_.process(...);
if (!result) {
  ctrl.warn(std::move(result.error()));
  co_yield {};
} else {
  co_yield result;
}
```

I'm leaving it be.

**Pull request merged**